recipes-extended/xen/blktap3/fix-encryption.patch

```diff
@@ -0,0 +1,169 @@
+Fix vhd-encryption failures caused by upstream blktap3 changes
+
+################################################################################
+LONG DESCRIPTION:
+################################################################################
+After upstreaming the OpenXT blktap encryption logic, several changes were
+merged into the upstream blktap repository that broke vhd encryption on OpenXT.
+These commits were 032df683d4d86502e2193337d9546e5b07429e77 and
+63e7784aaa05f8c2f96a09af50f6131af11f26b9
+A summary of the problems are as follows:
+
+ 1. The find key logic was if-def'd out. We now set this in blktap3.bb
+ 2. Because of 1, the redefinition of key causes a variable redefine error when
+    we turn on the if-def. Rename the var in function def to avoid this.
+ 3. 63e7784 only loads libblockcrypto if -E is set when calling tap-ctl-open.
+    Firstly, we always want to load the crypto support, so drop the dummy_load
+    routine. Secondly, we use tap-ctl-create, which was defaulting to NULL and 
+    0 for encryption and keysize since -E is only defined for tap-ctl-open.
+    This is also addressed by dropping the dummy load and if check
+ 4. findkey() sets keysize to 512 or 256, but an xts case deals with bytes,
+    so we divide by 8 before xts_aes_setkey()
+
+################################################################################
+CHANGELOG
+################################################################################
+Authors:
+Chris Rogers <rogersc@ainfosec.com>
+
+################################################################################
+REMOVAL
+################################################################################
+When upstreamed.
+
+################################################################################
+UPSTREAM PLAN
+################################################################################
+Should be upstreamed back to xapi project, but will take more time to update our
+implementation to fit their design changes. This needs to be in our tree ASAP
+so we patch here first, and upstream later.
+
+################################################################################
+INTERNAL DEPENDENCIES
+################################################################################
+
+################################################################################
+PATCHES
+################################################################################
+--- a/drivers/block-crypto.c
++++ b/drivers/block-crypto.c
+@@ -344,7 +344,7 @@ out:
+ #endif
+ 
+ int
+-vhd_open_crypto(vhd_context_t *vhd, const uint8_t *key, size_t key_bytes, const char *name)
++vhd_open_crypto(vhd_context_t *vhd, const uint8_t *keyin, size_t key_bytes, const char *name)
+ {
+ 	struct vhd_keyhash keyhash;
+ 	int err;
+@@ -402,7 +402,7 @@ vhd_open_crypto(vhd_context_t *vhd, cons
+ 		return err;
+ 	}
+ 
+-	xts_aes_setkey(vhd->xts_tfm, key, key_bytes);
++	xts_aes_setkey(vhd->xts_tfm, key, keysize / 8);
```

Comment:
As the keysize is pulled in within the #ifdef earlier, can we set key_bytes to equal keysize/8 within the #ifdef code? (or have another variable which is initialised to key_bytes and overridden to keysize/8 within the ifdef)

Reply:
(Also, IMO, keysize should be keysize_bits or keysize_bytes to be clear - as there are mismatches throughout)

Reply:
Yea the inconsistencies threw me off a bit when I was debugging.  I can do the variable renaming to note bits/bytes, that's not a problem.  Dividing by 8 after the return from ```chain_find_keyed_vhd``` will also drop the need for this hunk.